### PR TITLE
:tada: Support for images as fills

### PR DESCRIPTION
**Pull request by superalex**
https://tree.taiga.io/project/penpot/us/2983?milestone=360763

### CODE REVIEW

File: frontend/src/app/main/data/workspace.cljs [hunk 1]
```diff
@@ -1506,27 +1506,45 @@
...
-                                    :uri url
-                                    :response-type :blob})
-                       (rx/map :body)
-                       (rx/mapcat wapi/read-file-as-data-url)
-                       (rx/map #(assoc obj ::data %))
-                       (rx/take 1)))
+            (let [obj (maybe-translate obj objects parent-frame-id)
+                  ;; Texts can have different fills for pieces of the text
+                  fill-images-data   (->> (or (:position-data obj) [obj])
+                                          (mapv :fills)
```
Comment (superalex) [inline frontend/src/app/main/data/workspace.cljs]: filter & map

File: backend/src/app/tasks/file_gc.clj [hunk 2]
```diff
@@ -112,18 +112,23 @@
...
+                               fill-image-ids (->> (:fills obj)
+                                                   (mapv :fill-image)
+                                                   (mapv :id))
+                               stroke-image-ids (->> (:strokes obj)
+                                                     (mapv :stroke-image)
+                                                     (mapv :id))]
+                           (->> (into
+                                 [fill-image-id metadata-id]
+                                 (concat fill-image-ids stroke-image-ids))
+                               (filter some?))))))
```
Comment (niwinz) [inline backend/src/app/tasks/file_gc.clj]: You probably can replace this code with:

```clojure
(sequence
 (keep :id)
 (concat [(:fill-image obj)
          (:metadata obj)]
         (map :fill-image (:fills obj))
         (map :stroke-image (:strokes obj))))
```

File: frontend/src/app/main/ui/shapes/export.cljs [hunk 3]
```diff
@@ -286,32 +287,41 @@
...
      [:> "penpot:fills" #js {}
       (for [[index fill] (d/enumerate fills)]
-        [:> "penpot:fill"
-         #js {:penpot:fill-color          (if (some? (:fill-color-gradient fill))
-                                              (str/format "url(#%s)" (str "fill-color-gradient_" (mf/use-ctx muc/render-id) "_" index))
+        (let [fill-image-id (dm/str "fill-image-" (mf/use-ctx muc/render-id) "-" index)]
+          [:> "penpot:fill"
+           #js {:penpot:fill-color          (cond
+                                              (some? (:fill-color-gradient fill))
+                                              (str/format "url(#%s)" (str "fill-color-gradient-" (mf/use-ctx muc/render-id) "-" index))
```
Comment (niwinz) [inline frontend/src/app/main/ui/shapes/export.cljs]: str/ffmt & use-ctx should be on the top let

File: frontend/src/app/main/ui/shapes/export.cljs [hunk 4]
```diff
@@ -286,32 +287,41 @@
...
-              :penpot:stroke-color-ref-file (d/name (:stroke-color-ref-file stroke))
-              :penpot:stroke-color-ref-id   (d/name (:stroke-color-ref-id stroke))
-              :penpot:stroke-opacity        (d/name (:stroke-opacity stroke))
-              :penpot:stroke-style          (d/name (:stroke-style stroke))
-              :penpot:stroke-width          (d/name (:stroke-width stroke))
-              :penpot:stroke-alignment      (d/name (:stroke-alignment stroke))
-              :penpot:stroke-cap-start      (d/name (:stroke-cap-start stroke))
-              :penpot:stroke-cap-end        (d/name (:stroke-cap-end stroke))}])])))
-
+        (let [stroke-image-id (dm/str "stroke-image-" (mf/use-ctx muc/render-id) "-" index)]
```
Comment (niwinz) [inline frontend/src/app/main/ui/shapes/export.cljs]: use-ctx incorreclty used, should be declared on the toplevel let of the component

File: frontend/src/app/main/ui/shapes/fills.cljs [hunk 5]
```diff
@@ -36,21 +36,25 @@
...
 
                       (some? image)
-                          (cfg/resolve-file-media image))
+                      (cfg/resolve-file-media image))
 
-        embed       (embed/use-data-uris [uri])
+
+        uris (->>
+              (into [uri] (mapv #(when (:fill-image %) (cfg/resolve-file-media (:fill-image %))) fills))
+              (filter some?))
```
Comment (niwinz) [inline frontend/src/app/main/ui/shapes/fills.cljs]:
```clojure
(let [uris (into [uri]
                 (comp
                  (keep :fill-image)
                  (map cfg/resolve-file-media))
                 fills)]
  ...
  )

File: frontend/src/app/main/ui/shapes/fills.cljs [hunk 6]
```diff
@@ -36,21 +36,25 @@
...
+
+        embed       (embed/use-data-uris uris)
         transform   (gsh/transform-str shape)
 
-        ;; When tru e the image has not loaded yet
-        loading?    (and (some? uri)
-                         (not (contains? embed uri)))
+        ;; When true the image has not loaded yet
+        loading?    (not-any? #(contains? embed %) uris)
```
Comment (niwinz) [inline frontend/src/app/main/ui/shapes/fills.cljs]: Prefer `(partial contains? embed)`

File: frontend/src/app/worker/import.cljs [hunk 7]
```diff
@@ -317,36 +317,59 @@
...
+        (cip/has-stroke-images? node)
+        (cip/has-fill-images? node))
+    (let [name               (cip/get-image-name node)
+          has-image          (cip/has-image? node)
+          image-data         (cip/get-image-data node)
+          image-fill         (cip/get-image-fill node)
+          fill-images-data   (->> (cip/get-fill-images-data node)
+                                  (mapv #(assoc % :type :fill)))
+          stroke-images-data (->> (cip/get-stroke-images-data node)
+                                  (mapv #(assoc % :type :stroke)))
```
Comment (niwinz) [inline frontend/src/app/worker/import.cljs]: mapv -> map

File: frontend/src/app/worker/import.cljs [hunk 8]
```diff
@@ -379,7 +402,9 @@
              (rx/mapcat
               (fn [node]
                 (->> (resolve-media context file-id node)
-                     (rx/map (fn [result] [node result])))))
+                     (rx/map (fn [result]
+                               (js/console.log "pre-process-images" (clj->js result))
```
Comment (niwinz) [inline frontend/src/app/worker/import.cljs]: :fire:

**Pull request merged**